umbraco/Umbraco-CMS
# V14: Blueprint CRUD endpoints

## Details
- Added the following Blueprint endpoints:
  - Get blueprint by id - **GET** `/umbraco/management/api/v1/document-blueprint/{id}`
  - Delete blueprint by id - **DELETE** `/umbraco/management/api/v1/document-blueprint/{id}`
  - Update blueprint by id - **PUT** `/umbraco/management/api/v1/document-blueprint/{id}`
  - Create new blueprint - **POST** `/umbraco/management/api/v1/document-blueprint`
  - Create blueprint from existing content - **POST** `/umbraco/management/api/v1/document-blueprint/from-document`
- Introduced `IContentBlueprintEditingService` which will follow refactoring to clean up the shared implementation;
  - Tests will be added later.
- Introduced new Blueprint models and mappings;
  - Added base classes to reuse models between Documents and Blueprints.

> [!NOTE]  
> In both _Create_ endpoints, the `parent` object in the request models is not currently utilized; this will make it easier in case we introduce blueprint containers later on:
```json
  "parent": {
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6"
  }
```
---
**Bonus**: 
- Making **ContentControllerBase** abstract;
- Fix base inheritance case in `SchemaIdSelector`;
- Fixed a bug in document type mapping:
  - Mapping the incorrect key resulted in having the same id for both the blueprint and its document type in `GET
/umbraco/management/api/v1/tree/document-blueprint/root`.
- Fixed a bugs in `GET
/umbraco/management/api/v1/item/document-blueprint`:
  - Returning more entities than it was supposed to;
  - Mapping the Blueprint name instead of the document type name.

## Test
- Create Document types and Document in v13 + Content templates (blueprints);
- Migrate to this branch;
- Make sure the newly introduced CRUD endpoints work as expected;
- Test that both the Root and Item endpoints (in the **Document Blueprint** swagger group) return the correct data for blueprints.
